### PR TITLE
Remove jQuery as a dependency

### PR DESCRIPTION
**Pull request by milesj**
This PR removes jQuery as a dependency and replaces the outer/inner width/height methods with a single function that handles all scenarios. Will need to do further testing in browser and within unit tests to verify the numbers are correct.

### CODE REVIEW

File: src/components/DayPicker.jsx [hunk 1]
```diff
@@ -22,10 +21,11 @@ import { HORIZONTAL_ORIENTATION, VERTICAL_ORIENTATION } from '../constants';
 const CALENDAR_MONTH_WIDTH = 300;
 const DAY_PICKER_PADDING = 9;
 const MONTH_PADDING = 23;
-
 const PREV_TRANSITION = 'prev';
 const NEXT_TRANSITION = 'next';
 
+const slice = Array.prototype.slice;
```
Comment (majapw) [inline src/components/DayPicker.jsx]: This is a really weird thing to pull out, I feel.

File: src/components/DayPicker.jsx [hunk 2]
```diff
@@ -190,27 +242,31 @@ export default class DayPicker extends React.Component {
...
 
   adjustDayPickerHeight() {
-    const $transitionContainer = $(ReactDOM.findDOMNode(this.refs.transitionContainer));
-    const heights = $transitionContainer.find('.CalendarMonth[data-visible="true"]').get().map(
-      (month) => this.getMonthHeight($(month))
-    );
+    const transitionContainer = ReactDOM.findDOMNode(this.refs.transitionContainer);
+    const heights = [];
+
+    slice.call(transitionContainer.querySelectorAll('.CalendarMonth')).forEach((el) => {
```
Comment (majapw) [inline src/components/DayPicker.jsx]: I feel like this is not the most readable way of doing this?

File: src/components/DayPicker.jsx [hunk 3]
```diff
@@ -2,7 +2,6 @@ import React, { PropTypes } from 'react';
 import shallowCompare from 'react-addons-shallow-compare';
 import ReactDOM from 'react-dom';
 import moment from 'moment';
-import $ from 'jquery';
```
Comment (kesne) [inline src/components/DayPicker.jsx]: I love you.

File: src/components/DayPicker.jsx [hunk 4]
```diff
@@ -104,14 +104,63 @@ export default class DayPicker extends React.Component {
   }
 
   getMonthHeightByIndex(i) {
-    const $transitionContainer = $(ReactDOM.findDOMNode(this.refs.transitionContainer));
-    return this.getMonthHeight($transitionContainer.find('.CalendarMonth').eq(i));
+    return this.getMonthHeight(
+      ReactDOM.findDOMNode(this.refs.transitionContainer).querySelectorAll('.CalendarMonth')[i]
```
Comment (kesne) [inline src/components/DayPicker.jsx]: This looks like it's a dom ref, so you don't need to do findDOMNode.

File: src/components/DayPicker.jsx [hunk 5]
```diff
@@ -104,14 +104,63 @@ export default class DayPicker extends React.Component {
...
+    const caption = el.querySelector('.js-CalendarMonth__caption');
+    const grid = el.querySelector('.js-CalendarMonth__grid');
+
     // Need to separate out table children for FF
-    return $el.find('.js-CalendarMonth__caption').outerHeight(true) +
-      $el.find('.js-CalendarMonth__grid').height() + 1; // for border
+    return (
+      this.calculateDimension(caption, 'height', 'outer', true) +
+      this.calculateDimension(grid, 'height') +
+      1 // for border
```
Comment (kesne) [inline src/components/DayPicker.jsx]: Comments should go before lines. https://github.com/airbnb/javascript#comments--singleline

File: src/components/DayPicker.jsx (hunk 2 above)
Comment (kesne) [inline src/components/DayPicker.jsx]: You could use `Array.from` and then force people to polyfill.

Comment (kesne) [inline src/components/DayPicker.jsx]: Or pull a module to do this! 😀

Comment (majapw) [inline src/components/DayPicker.jsx]: I'm pretty against forcing people to polyfill. The whole point is for this to Just Work.

File: src/components/DayPicker.jsx [hunk 6]
```diff
@@ -190,27 +242,31 @@ export default class DayPicker extends React.Component {
...
+      }
+    });
 
     const newMonthHeight = Math.max(...heights) + MONTH_PADDING;
-    if (newMonthHeight !== $transitionContainer.height()) {
+
+    if (newMonthHeight !== this.calculateDimension(transitionContainer, 'height')) {
       this.monthHeight = newMonthHeight;
-      $transitionContainer.css('height', newMonthHeight);
+      transitionContainer.style.height = newMonthHeight;
```
Comment (razh) [inline src/components/DayPicker.jsx]: I think this needs a `px`.

File: src/components/DayPicker.jsx (hunk 2 above)
Comment (milesj) [inline src/components/DayPicker.jsx]: `Array.prototype.slice` is the way to do it. Thought I was going to be using it more, hence the reference at the top.

File: src/components/DayPicker.jsx (hunk 4 above)
Comment (milesj) [inline src/components/DayPicker.jsx]: This is necessary for React 0.14.

File: src/components/DayPicker.jsx (hunk 6 above)
Comment (milesj) [inline src/components/DayPicker.jsx]: Right, nice catch!

File: src/components/DayPicker.jsx [hunk 7]
```diff
@@ -190,27 +240,33 @@ export default class DayPicker extends React.Component {
...
-    const $transitionContainer = $(ReactDOM.findDOMNode(this.refs.transitionContainer));
-    const heights = $transitionContainer.find('.CalendarMonth[data-visible="true"]').get().map(
-      (month) => this.getMonthHeight($(month))
-    );
+    const transitionContainer = ReactDOM.findDOMNode(this.refs.transitionContainer);
+    const calendarMonths = transitionContainer.querySelectorAll('.CalendarMonth');
+    const heights = [];
+
+    // convert node list to array
+    Array.prototype.slice.call(calendarMonths).forEach((el) => {
```
Comment (ljharb) [inline src/components/DayPicker.jsx]: `Array.prototype.forEach.call` - no need to create the intermediate array.

Alternatively, `[...calendarMonths].forEach` but the former is faster.

File: package.json [hunk 8]
```diff
@@ -80,7 +81,6 @@
     "react-portal": "^2.2.1"
   },
   "peerDependencies": {
-    "jquery": ">=2.1",
```
Comment (ljharb) [inline package.json]: please also update the readme to remove jquery from the example generated `npm install` command

File: src/components/DayPicker.jsx (hunk 7 above)
Comment (milesj) [inline src/components/DayPicker.jsx]: Totally forgot about spreading. Will give that a shot (since it will probably compile down to Array prototype anyways).

Comment (ljharb) [inline src/components/DayPicker.jsx]: spreading will compile down to effectively a slice + forEach - forEach.call will just be the forEach (so i'd recommend doing that)

**Comment by milesj** [comment]
Some screenshots of the calculations in action. Compare the numbers in the console to the dimensions of the hovered element.

screen shot 2016-08-19 at 10 19 17 am
screen shot 2016-08-19 at 10 19 25 am
screen shot 2016-08-19 at 10 20 02 am

I'm not sure how to write unit tests for this as their is no DOM. There were no tests before for it anyways.

**Comment by majapw** [comment]
Yeah of my TODOS is to enable dom testing at some point which would help.
You could stub some things but it might be somewhat meaningless.

Do the height animations still work?

**Comment by majapw** [comment]
Also can we test cross-browser before we merge this in?

**Comment by milesj** [comment]
I've tested it on Mac FF/Chrome/Safari with no issues. Trying to test IE and mobile.

File: package.json [hunk 9]
```diff
@@ -70,6 +69,7 @@
     "react-addons-test-utils": "^15.2.1",
     "react-dom": "^15.2.1",
     "react-svg-loader": "^1.1.1",
+    "rimraf": "^2.5.4",
```
Comment (ianks) [inline package.json]: Why was this added?

**Comment by Daniel15** [comment]
> Trying to test IE and mobile.

I can test in IE + Edge for you if you provide a public URL where I can hit your fork! :smile:

File: package.json (hunk 9 above)
Comment (milesj) [inline package.json]: Because it's required to run `npm build`. It was missing.

**Comment by milesj** [comment]
@Daniel15 I won't be using a public URL. Was just going to be testing locally.

File: src/components/DayPicker.jsx (hunk 4 above)
Comment (kesne) [inline src/components/DayPicker.jsx]: Not quite: https://facebook.github.io/react/blog/2015/10/07/react-v0.14.html#dom-node-refs

**Comment by milesj** [comment]
I tested in IE11 and Edge with no issues. Storybook doesn't work in IE9/10, so can't verify.

Also, note to self, update this code. (I can't because of Windows line endings).

```
  calculateDimension(el, axis, borderBox = false, withMargin = false) {
    if (!el) {
      return 0;
    }

    const style = window.getComputedStyle(el);
    const axisStart = (axis === 'width') ? 'Left' : 'Top';
    const axisEnd = (axis === 'width') ? 'Right' : 'Bottom';

    // Includes border and padding
    let size = (axis === 'width') ? el.offsetWidth : el.offsetHeight;

    // Get the inner size
    if (!borderBox) {
      size -= (
        parseFloat(style[`padding${axisStart}`]) +
        parseFloat(style[`padding${axisEnd}`]) +
        parseFloat(style[`border${axisStart}Width`]) +
        parseFloat(style[`border${axisEnd}Width`])
      );
    }

    // Apply margin
    if (withMargin) {
      size += (
        parseFloat(style[`margin${axisStart}`]) +
        parseFloat(style[`margin${axisEnd}`])
      );
    }

    return size;
  }
```

File: src/components/DayPicker.jsx [hunk 10]
```diff
@@ -104,14 +102,63 @@ export default class DayPicker extends React.Component {
...
+      el.style[styleKey] = transformStyles[styleKey];
+    });
+  }
+
+  calculateDimension(el, axis, range = 'inner', withMargin = false) {
+    if (!el) {
+      return 0;
+    }
+
+    const style = window.getComputedStyle(el);
```
Comment (Rendez) [inline src/components/DayPicker.jsx]: Would it be an option to provide padding and border as inline styles to the component to avoid getting computed style?

Comment (ljharb) [inline src/components/DayPicker.jsx]: Computed style is about its actual size - padding and border just contribute to its desired size. User stylesheets, different font sizes, etc, could all effect the rendered dimensions.

Comment (Rendez) [inline src/components/DayPicker.jsx]: Sure, but if border and padding and margin are passed as props and used for inline css, then getBoundingClientRect() could be used instead, which is x100 faster

Comment (ljharb) [inline src/components/DayPicker.jsx]: Ah, I see what you mean. That's not a bad idea, but might require a bit too much coupling.

**Comment by milesj** [comment]
@ljharb @majapw Updated the code. Added some stub tests as well.

All I really need to do is test IE8/9/10, but not sure how, because storybook does not work in it.

**Comment by majapw** [comment]
@milesj So the best bet is probably to do this using airbnb.com at this point. We can `npm link` to this repo locally and specifically to this branch, and check to see if our homepage works as expected. I can probably do this later today or tomorrow, although it will probably require another PR first. Are you okay with holding off on this for a few more days? We can have this in a minor release then. 

I really don't want to merge this in and then after the fact discover that things are broken in those browsers.

**Comment by milesj** [comment]
Yeah, no problem. No rush to get this out right now.

**Comment by majapw** [comment]
So I tested on IE9, and it works as expected (no animations, but no errors). Same as before.

IE10 works well. :) Have you looked at mobile safari and mobile chrome?

**Comment by milesj** [comment]
Haven't tested mobile, but I don't foresee any issues. It's pretty straight forward.

I'll rebase tomorrow morning.

**Comment by majapw** [comment]
Awesome. :+1:

**Comment by milesj** [comment]
@majapw Rebased. Passed linting and tests locally.
